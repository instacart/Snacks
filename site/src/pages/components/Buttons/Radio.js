import React from 'react'
import Layout from '../../../components/Layout'
import Playground from '../../../components/Playground'
import {Title2, Title3} from '../../../components/Typography'

export default () => (
<Layout>

This component outputs a radio button with an optional label.

You can define whether the button starts in the pressed on unpressed state through the `isSelected` prop. In addition, you can change the selected/unselected state at any moment after the mount state by redifining the prop `isSelected`.

**Note:** Radio buttons are not supposed to be unselected once pressed. If you need that behavior, a button or a checkbox is a better solution.

<Title3>Label</Title3>
Associating a label to every radio button is very important for accessibility purposes. There are two ways to do it. The easiest is to enclose the text with the component: `<Radio>My label<Radio>`. The other option is to associate a label yourself:

<Playground>
  {`
<div style={{display: 'flex', flexDirection: 'column'}}>
  <div style={{display: 'flex', marginBottom: 20}}>
    <Radio id="radio1" isSelected />
    <label htmlFor="radio1" style={{marginLeft: 10}}>Label associated manually</label>
  </div>
  <Radio id="radio2">Label auto-generated</Radio>
</div>
`}</Playground>

<Title3>Groups</Title3>
To create a radio group where only one button can be selected [see here](#radiogroup).

<Title3>Styling</Title3>
You can pass an object to the `style` prop for styling the label, radio button and wrap element using Radium's structure:
<Playground>
  {`
<Radio
  id="radio3"
  wrapEl='span'
  style={{
    wrapEl: {padding: 10, border: '2px dotted lightblue'},
    button: {marginRight: 40},
    label: {color: 'salmon'},
  }}
>
  My Radio Button Label
</Radio>
`}</Playground>


<Title3>Callback</Title3>
You can pass a callback through the `onChange` prop:

<Playground>
  {`
function onChange(event, props) {
  alert(`This button is ${props.isSelected ? '' : 'not '}selected`)
}

<Radio id="radio4" onChange={onChange}>Click me!</Radio>
`}</Playground>

<Title3>Disabling</Title3>
You can disable a checkbox by declaring the `isDisabled` prop:

**Important!** If you attach a label to a Radio button manually, make sure the styleguide is followed for the disabled state.

<Playground>
  {`
function onChange(event, props) {
  alert(`This function will not be called`)
}

<Radio id="radio5" onChange={onChange} isDisabled>I am disabled</Radio>
`}</Playground>

</Layout>
)
