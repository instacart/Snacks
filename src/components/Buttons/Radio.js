import React from 'react'
import PropTypes from 'prop-types'
import RadioCheckboxBase from '../../base/RadioCheckboxBase'
import radioBaseSvg from '../../assets/radioBase.svg'
import radioSelectedSvg from '../../assets/radioSelected.svg'

const renderInputButton = (isSelected, style) => {
  const SvgComponent = isSelected ? radioSelectedSvg : radioBaseSvg
  return <SvgComponent aria-hidden="true" style={style} />
}

const Radio = props => {
  return <RadioCheckboxBase btnType="radio" renderInputButton={renderInputButton} {...props} />
}

Radio.propTypes = {
  aria: PropTypes.shape({
    label: PropTypes.string,
  }),
  children: PropTypes.node,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  isDisabled: PropTypes.bool,
  isSelected: PropTypes.bool,
  name: PropTypes.string,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  style: PropTypes.shape({
    button: PropTypes.object,
    label: PropTypes.object,
    wrapEl: PropTypes.object,
  }),
  value: PropTypes.string,
  wrapEl: PropTypes.string,
}

export default Radio
