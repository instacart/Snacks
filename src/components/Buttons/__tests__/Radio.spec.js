import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import sinon from 'sinon'
import Radio from '../Radio'

describe('Radio', () => {
  it('renders the correct button type', () => {
    const tree = renderer.create(<Radio id={1} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders the correct selected state', () => {
    const testCases = [
      { id: 1, isSelected: false },
      { id: 2, isSelected: true },
      { id: 2, isSelected: true, isDisabled: true },
    ]

    testCases.forEach(props => {
      const tree = renderer.create(<Radio {...props} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  it('disables the input when it is disabled', () => {
    const wrapper = mount(
      <Radio id="A" value="test" isSelected>
        Testing disabled
      </Radio>
    )
    const htmlBtn = wrapper
      .find('#A')
      .hostNodes()
      .instance()

    expect(htmlBtn.checked).toBe(true)
    expect(htmlBtn.disabled).toBe(false)

    wrapper.setProps({ isDisabled: true })

    expect(htmlBtn.checked).toBe(true)
    expect(htmlBtn.disabled).toBe(true)
  })

  it('generates a label if a child prop of text is supplied', () => {
    const tree = renderer.create(<Radio id={1}>Test label</Radio>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('incorporates user styles if passed', () => {
    const testCases = [
      {
        id: 1,
        style: {
          wrapEl: { background: 'pink' },
          button: { border: '2px solid purple' },
          label: { color: 'green' },
          inputBtn: { cursor: 'pointer' },
        },
      },
      {
        id: 2,
        style: undefined,
      },
    ]

    testCases.forEach(props => {
      const tree = renderer.create(<Radio {...props}>Test label</Radio>).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  it('calls the user-supplied onChange function', () => {
    const onChange = sinon.spy()
    const wrapper = mount(<Radio id={1} onChange={onChange} />)

    wrapper.find('input').simulate('change')
    expect(onChange.calledOnce).toBe(true)
  })

  it('calls the user-supplied onFocus function', () => {
    const onFocus = sinon.spy()
    const wrapper = mount(<Radio id={1} onFocus={onFocus} />)

    wrapper.find('input').simulate('focus')
    expect(onFocus.calledOnce).toBe(true)
  })

  it('calls the user-supplied onBlur function', () => {
    const onBlur = sinon.spy()
    const wrapper = mount(<Radio id={1} onBlur={onBlur} />)

    wrapper.find('input').simulate('blur')
    expect(onBlur.calledOnce).toBe(true)
  })
})
