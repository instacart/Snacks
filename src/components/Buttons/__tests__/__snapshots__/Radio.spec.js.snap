// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio generates a label if a child prop of text is supplied 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <test-file-stub
      data-radium={true}
      style={
        Object {
          "fill": "#43B02A",
          "height": 22,
          "width": 22,
        }
      }
    />
    <input
      aria-label={undefined}
      checked={false}
      data-radium={true}
      disabled={undefined}
      id={1}
      onChange={[Function]}
      style={
        Object {
          "MozAppearance": "none",
          "MsAppearance": "none",
          "WebkitAppearance": "none",
          "appearance": "none",
          "backgroundImage": "none",
          "height": 22,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "width": 22,
        }
      }
      type="radio"
      value={undefined}
    />
  </div>
  <label
    data-radium={true}
    htmlFor={1}
    style={
      Object {
        "marginLeft": 10,
      }
    }
  >
    Test label
  </label>
</div>
`;

exports[`Radio incorporates user styles if passed 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "background": "pink",
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "border": "2px solid purple",
        "position": "relative",
      }
    }
  >
    <test-file-stub
      data-radium={true}
      style={
        Object {
          "fill": "#43B02A",
          "height": 22,
          "width": 22,
        }
      }
    />
    <input
      aria-label={undefined}
      checked={false}
      data-radium={true}
      disabled={undefined}
      id={1}
      onChange={[Function]}
      style={
        Object {
          "MozAppearance": "none",
          "MsAppearance": "none",
          "WebkitAppearance": "none",
          "appearance": "none",
          "backgroundImage": "none",
          "cursor": "pointer",
          "height": 22,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "width": 22,
        }
      }
      type="radio"
      value={undefined}
    />
  </div>
  <label
    data-radium={true}
    htmlFor={1}
    style={
      Object {
        "color": "green",
        "marginLeft": 10,
      }
    }
  >
    Test label
  </label>
</div>
`;

exports[`Radio incorporates user styles if passed 2`] = `
<div
  data-radium={true}
  style={
    Object {
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <test-file-stub
      data-radium={true}
      style={
        Object {
          "fill": "#43B02A",
          "height": 22,
          "width": 22,
        }
      }
    />
    <input
      aria-label={undefined}
      checked={false}
      data-radium={true}
      disabled={undefined}
      id={2}
      onChange={[Function]}
      style={
        Object {
          "MozAppearance": "none",
          "MsAppearance": "none",
          "WebkitAppearance": "none",
          "appearance": "none",
          "backgroundImage": "none",
          "height": 22,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "width": 22,
        }
      }
      type="radio"
      value={undefined}
    />
  </div>
  <label
    data-radium={true}
    htmlFor={2}
    style={
      Object {
        "marginLeft": 10,
      }
    }
  >
    Test label
  </label>
</div>
`;

exports[`Radio renders the correct button type 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <test-file-stub
    data-radium={true}
    style={
      Object {
        "fill": "#43B02A",
        "height": 22,
        "width": 22,
      }
    }
  />
  <input
    aria-label={undefined}
    checked={false}
    data-radium={true}
    disabled={undefined}
    id={1}
    onChange={[Function]}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "backgroundImage": "none",
        "height": 22,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="radio"
    value={undefined}
  />
</div>
`;

exports[`Radio renders the correct selected state 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <test-file-stub
    data-radium={true}
    style={
      Object {
        "fill": "#43B02A",
        "height": 22,
        "width": 22,
      }
    }
  />
  <input
    aria-label={undefined}
    checked={false}
    data-radium={true}
    disabled={undefined}
    id={1}
    onChange={[Function]}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "backgroundImage": "none",
        "height": 22,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="radio"
    value={undefined}
  />
</div>
`;

exports[`Radio renders the correct selected state 2`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <test-file-stub
    data-radium={true}
    style={
      Object {
        "fill": "#43B02A",
        "height": 22,
        "width": 22,
      }
    }
  />
  <input
    aria-label={undefined}
    checked={true}
    data-radium={true}
    disabled={undefined}
    id={2}
    onChange={[Function]}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "backgroundImage": "none",
        "height": 22,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="radio"
    value={undefined}
  />
</div>
`;

exports[`Radio renders the correct selected state 3`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <test-file-stub
    data-radium={true}
    style={
      Object {
        "fill": "#43B02A",
        "height": 22,
        "width": 22,
      }
    }
  />
  <input
    aria-label={undefined}
    checked={false}
    data-radium={true}
    disabled={true}
    id={2}
    onChange={[Function]}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "backgroundImage": "none",
        "height": 22,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="radio"
    value={undefined}
  />
</div>
`;
